components/places/src/storage.rs

```diff
@@ -263,21 +270,314 @@ fn add_visit(db: &impl ConnExt,
+        Record(HistoryRecord),
+        Tombstone,
+    }
+
+    pub fn fetch_outgoing(db: &impl ConnExt,
+                          max_places: usize,
+                          max_visits: usize) -> Result<HashMap<SyncGuid, OutgoingInfo>> {
+        // Note that we want *all* "new" regardless of change counter,
+        // so that we do the right thing after a "reset".
+        let mut stmt = db.conn().prepare(&format!("
+            SELECT guid, url, id, title, hidden, typed, frecency,
+                visit_count_local, visit_count_remote,
+                last_visit_date_local, last_visit_date_remote,
+                sync_status, sync_change_counter
+            FROM moz_places
+            WHERE (sync_change_counter > 0 OR sync_status != {})
+            ORDER BY frecency DESC
+            LIMIT :max_places", (SyncStatus::Normal as u8)))?;
+        let mut visits = db.conn().prepare("
+            SELECT visit_date as date, visit_type as transition
+            FROM moz_historyvisits
+            WHERE place_id = :place_id
+            ORDER BY visit_date DESC
+            LIMIT :max_visits")?;
+        // tombstones
+        let mut tombstones_stmt = db.conn().prepare("
+            SELECT guid FROM moz_places_tombstones LIMIT :max_places")?;
+
+        let mut result: HashMap<SyncGuid, OutgoingInfo> = HashMap::new();
+
+        // We want to limit to 5000 places - tombstones are arguably the
+        // most important, so we fetch these first.
+        let ts_rows = tombstones_stmt.query_map_named(
+                            &[(":max_places", &(max_places as u32))],
+                            |row| SyncGuid(row.get::<_, String>("guid")))?;
+        for r in ts_rows {
+            let guid = r?;
+            trace!("outgoing tombstone {:?}", &guid);
+            result.insert(guid.clone(), OutgoingInfo::Tombstone);
+        }
+
+        // Max records is now limited by how many tombstones we found.
+        let max_places_left = max_places - result.len();
+
+        // We write info about the records we are updating to a temp table.
+        // While we could carry this around in memory, we'll need a temp table
+        // in `finish_outgoing` anyway, because we execute a `NOT IN` query
+        // there - which, in a worst-case scenario, is a very large `NOT IN`
+        // set.
+        db.execute_all(&["DROP TABLE IF EXISTS temp_sync_updated_meta",
+                         "CREATE TEMP TABLE temp_sync_updated_meta
+                          (id INTEGER PRIMARY KEY,
+                           change_delta INTEGER NOT NULL)
+                          WITHOUT ROWID"])?;
+
+        let insert_meta_sql = "
+            INSERT INTO temp_sync_updated_meta VALUES (:row_id, :change_delta)";
+
+        let rows = stmt.query_and_then_named(&[(":max_places",&(max_places_left as u32))],
+                                             PageInfo::from_row)?;
+        let mut ids_to_update = Vec::with_capacity(rows.size_hint().1.unwrap_or(500));
```

Comment:
nit: rusqlite row iterators don't give size hints, and in the case we don't know a size, I tend to avoid guessing the capacity (500 seems pretty high except for first syncs).